File: IF Quest/Em_java/Personagem.java
public abstract class Personagem{
    public abstract String habilidade();
    protected  String nome;
    protected int vida;
    protected int nivel;
    protected int forca;
    private Item[] inventario;
    private int quantidade_item;
    public Personagem(String nome, int vida, int nivel, int forca){
        this.nome = nome;
        this.nivel = nivel;
        this.setVida(vida);
        this.setForca(forca);
        inventario = new Item[10];
        quantidade_item = 0;
    }
    public String getNome(){return this.nome;}
    public int getVida(){return this.vida;}
    public int getNivel(){return this.nivel;}
    public int getForca(){return this.forca;}

    public void pegar(Item item){
        if(item == null){ 
            System.out.println("ERRO, item nulo\n");
            return;
        }
        if(quantidade_item < inventario.length){
            inventario[quantidade_item] = item;
            quantidade_item++;
        }else System.out.println("Inventario cheio");
    }
    public void setNome(String nome){
        if (nome == null || nome.isEmpty()){
            throw new IllegalArgumentException("Nome inválido: " + nome);
        }
        this.nome = nome;
    }
    public void setNivel(int nivel){
        if (nivel < 1){
            System.out.println("Nivel nao pode ser menor que 1");
        }else{
            this.nivel = nivel;
        }
    }
    public void setVida(int vida){
        if (vida > 200)throw new IllegalArgumentException("Vida nao pode ser negativa: " + vida);
        if (vida < 0) this.vida = 0;
        this.vida = vida;
    }
    public void setForca(int forca){
        if (forca < 0)throw new IllegalArgumentException("Forca nao pode ser negativa: " + forca);
        this.forca = forca;
    }
    public void atacar(Personagem alvo) throws SemMana{
        if (alvo != null) {
            int danoTotal = this.forca;
            for (int i = 0; i < quantidade_item; i++){
                danoTotal += this.inventario[i].getBonus();
            }
            alvo.setVida(alvo.getVida() - danoTotal);
            System.out.println(this.nome + " atacou " + alvo.getNome() + " causando " + danoTotal + " de dano total!");
            } else {
                System.out.println("ERRO: Nao ha alvo para atacar.\n");
        }
    }
    public void GolpeEspecial(String habilidade, int custoForca) throws SemForca {
        if (this.forca < custoForca) {
            throw new SemForca(this.forca);
        }
        this.forca -= custoForca; // Consome a força do personagem
        System.out.println(this.nome + " usou o golpe especial '" + habilidade + "' gastando " + custoForca + " de força!");
    }
    public void ficha(){
        System.out.println("Nome: " + this.nome);
        System.out.println("Vida: " + this.vida);
        System.out.println("Nivel: " + this.nivel);
        System.out.println("Forca: " + this.forca);
        System.out.println("Habilidade: " + habilidade());
        System.out.println("Inventario: ");
        if(this.quantidade_item == 0){
            System.out.println("Inventario vazio");
        }else{
            for(int i = 0; i < quantidade_item; i++){
                System.out.println(inventario[i].Descricao());
            }
        }
    }
}

File: IF Quest/Em_python/Personagem.py
from abc import ABC, abstractmethod
from Item import Item
from SemForca import SemForca
from SemMana import SemMana
class Personagem(ABC):
    def __init__(self, nome: str, vida: int, nivel: int, forca: int):
        self.nome = nome
        self.nivel = nivel
        self.vida = vida
        self.forca = forca
        self._inventario = [None] * 10
        self._quantidade_item = 0
    @abstractmethod
    def habilidade(self) -> str: 
        pass
    @property
    def nome(self):
        return self._nome
    @nome.setter
    def nome(self, nome):
        if nome is None or nome.strip() == "":
            raise ValueError(f"Nome inválido: {nome}")
        self._nome = nome
    @property
    def vida(self): 
        return self._vida
    @vida.setter
    def vida(self, vida):
        if vida > 200:
            raise ValueError(f"Vida nao pode ser maior que 200: {vida}")
        if vida < 0:
            self._vida = 0
        else:
            self._vida = vida
    @property
    def nivel(self):    
        return self._nivel
    @nivel.setter
    def nivel(self, nivel):
        if nivel < 1:
            print("Nivel nao pode ser menor que 1")
        else:
            self._nivel = nivel
    @property
    def forca(self):
        return self._forca
    @forca.setter
    def forca(self, forca):
        if forca < 0:
            raise ValueError(f"Forca nao pode ser negativa: {forca}")
        self._forca = forca
    def pegar(self, item: Item):
        if item is None: 
            print("ERRO, item nulo\n")
            return
        if self._quantidade_item < len(self._inventario):
            self._inventario[self._quantidade_item] = item
            self._quantidade_item += 1
        else: 
            print("Inventario cheio")
    def atacar(self, alvo=None):
        if alvo is not None:
            danoTotal = self.forca
            for i in range(self._quantidade_item):
                danoTotal += self._inventario[i].bonus
            alvo.vida = alvo.vida - danoTotal
            print(f"{self.nome} atacou {alvo.nome} causando {danoTotal} de dano total!")
        else:
            print("ERRO: Nao ha alvo para atacar.\n")
    def GolpeEspecial(self, habilidade: str, custoForca: int):
        if self.forca < custoForca:
            raise SemForca(self.forca)
        self.forca -= custoForca 
        print(f"{self.nome} usou o golpe especial '{habilidade}' gastando {custoForca} de força!")
    def ficha(self):
        print(f"Nome: {self.nome}")
        print(f"Vida: {self.vida}")
        print(f"Nivel: {self.nivel}")
        print(f"Forca: {self.forca}")
        if hasattr(self, '_mana') and self._mana < 10:
            print("Habilidade: Indisponível (Sem Mana)")
        else:
            if self.__class__.__name__ == "Mago":
                print("Habilidade: Magia_de_raio")
            elif self.__class__.__name__ == "Guerreiro":
                print("Habilidade: corte_duplo")
            else:
                print("Habilidade: Nenhuma")   
        print("Inventario: ")
        if self._quantidade_item == 0:
            print("Inventario vazio")
        else:
            for i in range(self._quantidade_item):
                print(self._inventario[i].descricao())
